File: src/01_Foundations/Variables_DataTypes.cpp
/**
 * Level 1: Foundations & Syntax
 * File: Variables_DataTypes.cpp
 * Topic: Primitive types, sizes, signed vs unsigned, and integer overflow traps.
 */

#include <iostream>
#include <climits>

int main() {
    int val = 2147483647;            // 4 Bytes: -2^31 to 2^31 - 1
    long long largeVal = 1e18;       // 8 Bytes: 64-bit integer (CRITICAL FOR DSA SUMS!)
    double pi = 3.141592653589793;   // 8 Bytes double precision
    char grade = 'A';                // 1 Byte ASCII
    bool flag = true;                // 1 Byte boolean

    std::cout << "int size: " << sizeof(int) << " bytes\n";
    std::cout << "long long size: " << sizeof(long long) << " bytes\n";
    std::cout << "Max Int: " << val << "\n";
    
    // Demonstrate overflow:
    int overflowed = val + 1;
    std::cout << "Overflowed Int (val + 1): " << overflowed << " (Wrapped around to negative!)\n";

    return 0;
}

File: src/01_Foundations/Variables_DataTypes.java
/**
 * Level 1: Foundations & Syntax
 * File: Variables_DataTypes.java
 * Topic: Primitive Types, Precision, Signed limits, Wrapper Classes
 */

public class Variables_DataTypes {
    public static void main(String[] args) {
        int val = Integer.MAX_VALUE;         // 4 Bytes: 2147483647
        long largeVal = 1000000000000000000L;// 8 Bytes: 64-bit integer ('L' required)
        double pi = 3.141592653589793;       // 8 Bytes double precision
        char grade = 'A';                    // 2 Bytes Unicode
        boolean flag = true;                 // 1 Byte boolean

        System.out.println("Max Int: " + val);
        System.out.println("Max Long: " + Long.MAX_VALUE);
        
        // Demonstrate overflow:
        int overflowed = val + 1;
        System.out.println("Overflowed Int (val + 1): " + overflowed + " (Two's complement wrap)");
    }
}
